graphql/src/store/query.rs

```diff
@@ -2,126 +2,182 @@ use graph::prelude::*;
                     GreaterOrEqual => StoreFilter::GreaterOrEqual(attribute, store_value),
                     LessOrEqual => StoreFilter::LessOrEqual(attribute, store_value),
-                    In => StoreFilter::In(attribute, list_values(store_value)),
-                    NotIn => StoreFilter::NotIn(attribute, list_values(store_value)),
+                    In => StoreFilter::In(attribute, list_values(store_value, "_in")?),
+                    NotIn => StoreFilter::NotIn(attribute, list_values(store_value, "_not_in")?),
                     Contains => StoreFilter::Contains(attribute, store_value),
                     NotContains => StoreFilter::NotContains(attribute, store_value),
                     StartsWith => StoreFilter::StartsWith(attribute, store_value),
                     NotStartsWith => StoreFilter::NotStartsWith(attribute, store_value),
                     EndsWith => StoreFilter::EndsWith(attribute, store_value),
                     NotEndsWith => StoreFilter::NotEndsWith(attribute, store_value),
                     Equal => StoreFilter::Equal(attribute, store_value),
-                }
-            }).collect::<Vec<StoreFilter>>(),
-    )
+                })
+            }).collect::<Result<Vec<StoreFilter>, QueryExecutionError>>()?
+    })))
 }
 
 /// Parses a list of GraphQL values into a vector of entity attribute values.
-fn list_values(value: Value) -> Vec<Value> {
+fn list_values(value: Value, filter_type: &str) -> Result<Vec<Value>, QueryExecutionError> {
     match value {
-        Value::List(values) => values,
-        _ => panic!("value is not a list"),
+        Value::List(ref values) if !values.is_empty() => {
+            // Check that all values in list are of the same type
+            let root_discriminant = discriminant(&values[0]);
+            values
+                .into_iter()
+                .map(|value| {
+                    let current_discriminant = discriminant(value);
+                    if root_discriminant == current_discriminant {
+                        Ok(value.clone())
+                    } else {
+                        Err(QueryExecutionError::ListTypesError(filter_type.to_string()))
+                    }
+                }).collect::<Result<Vec<_>, _>>()
+        }
+        Value::List(ref values) if values.is_empty() => Ok(vec![]),
+        _ => Err(QueryExecutionError::ListFilterError(
+            filter_type.to_string(),
+        )),
     }
 }
 
 /// Parses GraphQL arguments into an attribute name to order by, if present.
-fn build_order_by(arguments: &HashMap<&q::Name, q::Value>) -> Option<String> {
-    arguments
+fn build_order_by(
+    arguments: &HashMap<&q::Name, q::Value>,
+) -> Result<Option<String>, QueryExecutionError> {
+    Ok(arguments
         .get(&"orderBy".to_string())
         .and_then(|value| match value {
             q::Value::Enum(name) => Some(name.to_owned()),
             _ => None,
-        })
+        }))
```

Comment:
The formatting here strikes me as being off: shouldn't the closing `)` be on its own line, intended like `Ok`? I mean if `rustfmt` does it this way, I'll shut up 😉

Reply:
I agree, it makes sense to place the closing `)` at the same indent level as the opening `Ok(`...but this is how `rustfmt` likes it these days. 🤷